BoostTestAdapter/Boost/Runner/BoostTestRunnerBase.cs

```diff
@@ -136,11 +136,24 @@ private static int MonitorProcess(Process process, int timeout)
                 throw new TimeoutException(timeout);
             }
 
-            return process.ExitCode;
+            try
+            {
+                return process.ExitCode;
+            }
+            catch (InvalidOperationException)
+            {
+                // This is a common scenario when attempting to request the exit code
```

Comment:
Could you tell us what are the repro steps to get into that state?

I see in [Process.ExitCode](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.exitcode?view=netframework-4.7.2) docs that the exception can be thrown if the process has not exited (which should not be the case since there is a check for `process.HasExited` above) or its handle is not valid. Can the latter case be checked for with [Process.HandleCount](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.handlecount?view=netframework-4.7.2)?

Reply:
It is possible to reproduce this with Process instances which originate from DebugFrameworkExecutionContext.cs. The provisioned instances are merely references acquired from process IDs which are returned from the Visual Studio framework.

It might be possible to test for Process.HandleCount but I need time to verify it.

Reply:
I looked into trying to use Process.HandleCount but accessing this property throws an exception during either of the discovery, test execution or debug test execution phases.

Reply:
Thanks for trying this.

I have now spent some more time researching this and it appears to me that the cleanest way to handle this would be to enable [`Process.EnableRaisingEvents`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.enableraisingevents?redirectedfrom=MSDN&view=netframework-4.7.2#System_Diagnostics_Process_EnableRaisingEvents) and handle [`Process.Exited`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.exited?view=netframework-4.7.2). I was able to prototype a quick solution using these two and confirm it working. I guess the only complication would be synchronizing between the event handler and `MonitorProcess`, but even that should not be too difficult.

What do you think of switching to such approach?

Reply:
Tried the following prototype with similar exceptions being thrown in cases when the process has already terminated and the `executionContext` is of type `DebugFrameworkExecutionContext`:

```c#
public virtual int Execute(BoostTestRunnerCommandLineArgs args, BoostTestRunnerSettings settings, IProcessExecutionContext executionContext)
{
	Utility.Code.Require(settings, "settings");
	Utility.Code.Require(executionContext, "executionContext");

	var run = new TaskCompletionSource<int>();
	var task = run.Task;
	
	using (Process process = executionContext.LaunchProcess(GetExecutionContextArgs(args, settings)))
	{
		// process.WaitForExit();

		process.Exited += (object sender, EventArgs e) =>
		{
			run.SetResult(process.ExitCode);
		};

		try
		{
			process.EnableRaisingEvents = true;
		}
		catch (Exception ex)
		{
			// InvalidOperationException thrown if executionContext is
			// of type DebugFrameworkExecutionContext and the process
			// has already exited

			run.SetException(ex);
		}

		if (!task.Wait(settings.Timeout))
		{
			KillProcessIncludingChildren(process);
			throw new TimeoutException(settings.Timeout);
		}
	}

	return task.Result;
}
```

IMHO, comparing this approach to the one in the PR, the additional synchronisation code seems like extra overhead, especially since we still need to cater for the same exception. I do think that this approach may allow for an `async` interface, but this will require extensive work and API changes and might not be beneficial since we will still need to wait on the task in some other point in the adapter.

Reply:
I agree it is slightly more complicated, and add small overhead, but it also gives the correct exit code in more cases (i.e. unless the debugee exits too quickly). Let's integrate the simpler approach you have in the PR now, and leave the alternative as a future consideration, should you want to follow up on that later.